Support adding a domain in the purchase flow

gcsecsey opened the pull request:
#### Prerequisites

Before merging this PR:
- [X] We should decide where to route the `Learn more` link [related convo](https://github.com/Automattic/wp-calypso/issues/60444#issuecomment-1033944420)

#### Changes proposed in this Pull Request

* Adds a new dialog
* The dialog prompts the user to set their primary domain to a custom one if they have one or buy a custom domain
* The dialog is shown if the site's primary domain is a subdomain

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

The slug of Yoast Premium currently differ in the Matkeplace and WPCom REST API responses, but there's already [a diff to resolve that](D74410-code). The testing steps below will work once the slugs are updated:

**Displaying the dialog if there's no custom domain**
* switch to a site that does not have a custom domain
* visit a plugin that requires a custom primary domain (eg. `/plugins/wordpress-seo-premium`)
* click on the install button
* verify that the dialog with this text is displayed:
  <img width="498" alt="image" src="https://user-images.githubusercontent.com/11555574/153243144-9d38a8f7-893b-4648-b0f0-b8f130c7aa01.png">

**Displaying the dialog for non-primary custom domain**
* switch to a site that has a custom domain
* set the primary domain to be **other than the custom domain**
* visit a plugin that requires a custom primary domain (eg. `/plugins/wordpress-seo-premium`)
* click on the install button
* verify that the dialog with this text is displayed:
  <img width="499" alt="image" src="https://user-images.githubusercontent.com/11555574/153433397-af7c5afa-e818-4a97-88fa-f31aba966395.png">

**Not displaying the dialog for custom primary domain**
* switch to a site that has a custom domain (or custom subdomain)
* visit a plugin that requires a custom primary domain (eg. `/plugins/wordpress-seo-premium`)
* click on the install button
* verify that the dialog is not displayed

**Adding a domain**
* click `Manage domains`
* add a domain to the cart
* navigate back to the plugin page
* click install and choose `Install Yoast Premium`
* verify that the cart contains both the domain and the plugin

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #60444

gcsecsey committed:
[WIP] add selectors, create dialog

gcsecsey committed:
update styling

gcsecsey committed:
update styling

gcsecsey committed:
use default dialog styles for the custom domain dialog

gcsecsey committed:
show dialog based on the plugin's `required_primary_domain` field

a8ci18n commented on client/my-sites/plugins/plugin-custom-domain-dialog/index.tsx:
```diff
@@ -0,0 +1,75 @@
+	const buttons = [
+		{
+			action: 'learn-more',
+			label: translate( 'Learn more', { context: 'button label' } ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Learn more', { context: 'logged-out-homepage'} )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/domain-in-marketplace-purchase-flow)


:information_source: This string already exists with the following **contexts**:
- `null` (no context)
- `CMP banner`
- `learn more link to failed payment retry documentation`
- `learn more link to subscription emails documentation`
- `learn more link to subscription reports documentation`

 Would it make sense to reuse one of the above?

a8ci18n commented on client/my-sites/plugins/plugin-custom-domain-dialog/index.tsx:
```diff
@@ -0,0 +1,75 @@
+		{
+			action: 'manage-domains',
+			href: '/domains/manage/' + selectedSiteUrl,
+			label: translate( 'Manage domains', { context: 'button label' } ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **27** times:
`translate( 'Manage domains' )` *ES Score: 14*


:information_source: This string already exists without a **context**. Only add a context if the meaning of the string is very specific.

a8ci18n commented on client/my-sites/plugins/plugin-custom-domain-dialog/index.tsx:
```diff
@@ -0,0 +1,75 @@
+		},
+		{
+			action: 'install-plugin',
+			label: translate( 'Install %(pluginName)s', {
```
:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

gcsecsey committed:
translation fixes

gcsecsey committed:
fetch `requirements` key with WPCom marketplace products

cpapazoglou commented on client/my-sites/plugins/plugin-custom-domain-dialog/index.tsx:
```diff
@@ -0,0 +1,74 @@
+
+	const buttons = [
+		{
+			action: 'learn-more',
+			label: translate( 'Learn more' ),
```
Let's use `https://wordpress.com/support/domains/#set-a-primary-address` for the time being, instead of shipping a button that does nothing.

cpapazoglou reviewed (changes requested):
I've got a site with a custom domain set as primary but I still get the popup 
![CleanShot 2022-02-10 at 13 14 10@2x](https://user-images.githubusercontent.com/12430020/153396038-b923363b-4f96-484d-9b01-f110e265cfe9.png)

in this case, we shouldn't show the popup at all.

gcsecsey committed:
don't show dialog to custom subdomains

gcsecsey commented:
> in this case, we shouldn't show the popup at all.

Thanks for adding this @cpapazoglou! I've updated the condition to check for WPCom domains and subdomains, now it should work as expected. I've also updated the testing instructions.

cpapazoglou commented on client/my-sites/plugins/plugin-custom-domain-dialog/index.tsx:
```diff
@@ -0,0 +1,75 @@
+	const buttons = [
+		{
+			action: 'learn-more',
+			href: '/support/domains/#set-a-primary-address',
```
We should use a localised absolute link here

```suggestion
			href: localizeUrl( 'https://wordpress.com/support/domains/#set-a-primary-address' ),
```

cpapazoglou commented on client/my-sites/plugins/plugin-details-CTA/index.jsx:
```diff
@@ -217,6 +221,15 @@ const CTAButton = ( {
+	);
+
+	const pluginRequiresCustomPrimaryDomain =
+		( primaryDomain?.isWpcomDomain || primaryDomain?.isWpcomStagingDomain ) &&
```
`isWpcomDomain` should be `isWPCOMDomain` to match the case. Currently this is failing for simple sites.

gcsecsey replied:
Thanks for pointing this out! 🎉

cpapazoglou reviewed (dismissed):
I added some more comments. If we address those we should be good to 🚀 !

gcsecsey committed:
fix property name case, localize URL

gcsecsey commented:
The results of the **Translate (Web app)** job can be ignored for now, [related convo.](p1644491378696399/1644489911.531449-slack-C02JPCHEKDY)

cpapazoglou reviewed (approved):
LGTM

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7141825

Thank you @gcsecsey for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7141825).
